shared/domains/chelonia/utils.js

```diff
@@ -0,0 +1,151 @@
+          // and it's possible that an error here shouldn't necessarily break the entire
+          // contract. For example, in some situations we might read a contract as
+          // read-only and not have the key to write to it.
+        }
+      }
+    }
+
+    // Is this a an invite key? If so, run logic for invite keys and invitation
+    // accounting
+    if (key.name.startsWith('#inviteKey-')) {
+      if (!state._vm.invites) this.config.reactiveSet(state._vm, 'invites', Object.create(null))
+      this.config.reactiveSet(state._vm.invites, key.id, {
+        status: INVITE_STATUS.VALID,
+        initialQuantity: key.meta.quantity,
+        quantity: key.meta.quantity,
+        expires: key.meta.expires,
+        inviteSecret: decryptedKey || state._volatile?.keys?.[key.id],
+        responses: []
+      })
+    }
+
+    // Is this KEY operation the result of requesting keys for another contract?
+    console.log(['@@@@@KAP', key.meta?.keyRequest, findSuitableSecretKeyId(state, [GIMessage.OP_KEY_ADD], ['sig']), contractID])
+    if (key.meta?.keyRequest && findSuitableSecretKeyId(state, [GIMessage.OP_KEY_ADD], ['sig'])) {
+      const { id, contractID: keyRequestContractID, outerKeyId } = key.meta?.keyRequest
+
+      const rootState = sbp(this.config.stateSelector)
+
+      // Are we subscribed to this contract?
+      // If we are not subscribed to the contract, we don't set pendingKeyRequests because we don't need that contract's state
+      // Setting pendingKeyRequests in these cases could result in issues
+      // when a corresponding OP_KEY_SHARE is received, which could trigger subscribing to this previously unsubscribed to contract
+      if (keyRequestContractID) {
+        if (!rootState[keyRequestContractID]) {
+          this.config.reactiveSet(rootState, keyRequestContractID, { _volatile: { pendingKeyRequests: [] } })
+        } else if (!rootState[keyRequestContractID]._volatile) {
+          this.config.reactiveSet(rootState[keyRequestContractID], '_volatile', { pendingKeyRequests: [] })
+        } else if (!rootState[keyRequestContractID]._volatile.pendingKeyRequests) {
+          this.config.reactiveSet(rootState[keyRequestContractID]._volatile, 'pendingKeyRequests', [])
+        }
+
+        // Mark the contract for which keys were requested as pending keys
+        rootState[keyRequestContractID]._volatile.pendingKeyRequests.push({ id, outerKeyId })
+
+        this.setPostSyncOp(contractID, 'pending-keys-for-' + keyRequestContractID, ['okTurtles.events/emit', CONTRACT_IS_PENDING_KEY_REQUESTS, { contractID: keyRequestContractID }])
+      }
+    }
+
+    // Is this a foreign key in this contract? If so, flag it in the
+    // other contract to mirror operations
+    if (key.foreignKey) {
+      const fkUrl = new URL(key.foreignKey)
+      const foreignContract = fkUrl.pathname
+      const foreignKeyName = fkUrl.searchParams.get('keyName')
```

Comment:
As a user-provided parameter, can this be abused in any way?

Reply:
I'm not sure that this is 'user-provided' any more than the rest of the parameters, but I guess it would result in perhaps syncing the wrong key at worst.